### PR TITLE
Support `std::semaphore` and typed `std::mailbox`

### PR DESCRIPTION
**Pull request by kbieganski**
Adds support for `std::semaphore` and strictly-typed `std::mailbox`.

~~There are two main things that need some work:~~
* ~~Using these classes requires passing the `--std` flag. Not the end of the world, but it would be nice to be able to use them out of the box.~~
* ~~The checks for methods supported with `--timing` only take the class name into account, but they should check if it's from the actual built-in package.~~

### CODE REVIEW

File: bin/verilator [hunk 1]
```diff
@@ -398,6 +398,7 @@ detailed descriptions of these arguments.
     --no-skip-identical         Disable skipping identical output
     --stats                     Create statistics file
     --stats-vars                Provide statistics on variables
+    --std                       Verilate SystemVerilog's std package
```
Comment (wsnyder) [inline bin/verilator:401]: As you also noted, I don't want users to have to specify this.

My thinking was the parser would note a mailbox/std::mailbox/semaphore/std::semaphore went by, hack up the symbol tree, then after all other sources are done read the std.sv.

File: test_regress/t/t_semaphore.v [hunk 2]
```diff
@@ -16,11 +16,10 @@
  `define SEMAPHORE_T semaphore
 `endif
 
-module t(/*AUTOARG*/);
+module t_semaphore(/*AUTOARG*/);
```
Comment (wsnyder) [inline test_regress/t/t_semaphore.v:19]: Tests should be "module t", not sure why this change is needed.

File: test_regress/t/t_semaphore_class.v [hunk 3]
```diff
@@ -1,33 +0,0 @@
-// DESCRIPTION: Verilator: Verilog Test module
-//
-// This file ONLY is placed under the Creative Commons Public Domain, for
-// any use, without warranty, 2022 by Wilson Snyder.
-// SPDX-License-Identifier: CC0-1.0
-
-class semaphore_cls;
```
Comment (wsnyder) [inline test_regress/t/t_semaphore_class.v:7]: I'd like this test and t_mailbox_class to remain, it's harmess and users are allowed to do what they want.

**Review by wsnyder — CHANGES_REQUESTED** [review]
Please add an error if the user tries to do a "package std;" themselves.

File: bin/verilator (hunk 1 above)
Comment (kbieganski) [reply bin/verilator:401]: Did something like this.

File: test_regress/t/t_semaphore.v (hunk 2 above)
Comment (kbieganski) [reply test_regress/t/t_semaphore.v]: I want to do `-Wall` to catch any warnings from `std.sv`, but the module name would trigger `DECLFILENAME`. Added a `lint_off` instead.

Comment (wsnyder) [reply test_regress/t/t_semaphore.v]: FYI, The default for tests is Wall Wno-DECLFILENAME.

Comment (kbieganski) [reply test_regress/t/t_semaphore.v]: Yes, but I wanted `DECLFILENAME` to still take effect in `std.sv`.

File: include/std.sv [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        function new(int bound = 0);
+            m_bound = bound;
+        endfunction
+
+        function int num();
+            return m_queue.size();
+        endfunction
+
+        task put(T message);
+    `ifdef VERILATOR_TIMING
```
Comment (wsnyder) [inline include/std.sv:31]: Perhaps move ifdef around entire mailbox/semaphore, and have the parser error if --timing isn't used?

Code that needs a mailbox/semaphore just won't work without --timing, right?

File: src/V3ParseImp.cpp [hunk 5]
```diff
@@ -528,10 +528,13 @@ void V3ParseImp::tokenPipelineSym() {
                     token = yaID__ETC;
                 }
             }
+        } else if (token == yaID__CC && *(yylval.strp) == "std") {  // IEEE-standard package
+            v3Global.setUsesStdPackage();
```
Comment (wsnyder) [inline src/V3ParseImp.cpp:532]: Will this break if the user does "foo::std" where std is a class inside Foo package, this I believe is legal.  Add a test?

File: include/std.sv (hunk 4 above)
Comment (kbieganski) [reply include/std.sv:31]: Only the methods that block. I don't think these classes are very useful without those, but strictly speaking all the non-blocking methods should work without timing.

File: src/V3ParseImp.cpp (hunk 5 above)
Comment (kbieganski) [reply src/V3ParseImp.cpp:532]: It will "break" as in it will make us parse `std.sv`, but it'll still verilate. I think it's a rare enough case that it should be fine?

Comment (wsnyder) [reply src/V3ParseImp.cpp:532]: It will skip the other if statement branches below.  Maybe there's nothing harmful in that, but still seems worthy of a test to make sure.

Comment (kbieganski) [reply src/V3ParseImp.cpp:532]: Ah, I see. I'll add a test.

But the `else` should be executed in that `foo::std` case. Would something like:

```cpp
} else {  // Not found
    yylval.scp = nullptr;
    if (token == yaID__CC) {
        if (m_lexPrevToken != yaID__CC && *(yylval.strp) == "std") {
            v3Global.setUsesStdPackage();
        } else if (!v3Global.opt.bboxUnsup()) {
```
make sense? Seems to be working.

Comment (wsnyder) [reply src/V3ParseImp.cpp:532]: That seems reasonable.  All accesses should be a std:: or an import, both would be of a ID_CC.

Comment (kbieganski) [reply src/V3ParseImp.cpp]: It turned out to be more complicated. The tests dindn't cover `std::` usage properly.

File: include/std.sv (hunk 4 above)
Comment (kbieganski) [reply include/std.sv:31]: If we ifdef entire mailbox and semaphore, we'll still want a better error message that says what the problem is. E.g. libc++ >= 14 has an ifdef that checks if one is using C++20 around coroutine types, even if `-fcoroutines-ts` is passed, which is just confusing.

Comment (wsnyder) [reply include/std.sv:31]: Ok, leave it.

**Comment by wsnyder** [comment]
Is this good to go?  It's still in a changes requested mode - which is perhaps something I just need to clear.

**Comment by kbieganski** [comment]
I think it's good to go. Not sure why it still shows "Changes requested" (maybe it needs to be re-reviewed). Rebased it.

**Review by wsnyder — APPROVED** [review]
Looks good, feel free to squash merge once the tests pass.
